src/fcntl.rs

```diff
@@ -629,6 +632,102 @@ pub fn flock(fd: RawFd, arg: FlockArg) -> Result<()> {
 
     Errno::result(res).map(drop)
 }
+
+/// Represents valid types for flock.
+///
+/// # Safety
+/// `T` must be `!Clone`.
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+pub unsafe trait Flockable: AsRawFd {}
+
+/// Represents an owned flock, which unlocks on drop.
+///
+/// See flock(2) for details on locking semantics.
+// `ManuallyDrop` is necessary to circumvent move out of `Drop` type error.
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+#[derive(Debug)]
+pub struct Flock<T: Flockable>(ManuallyDrop<T>);
+
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> Drop for Flock<T> {
+    fn drop(&mut self) {
+        // Result is ignored because flock has no documented failure cases.
+        _ = flock(self.0.as_raw_fd(), FlockArg::Unlock);
+    }
+}
+
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> Deref for Flock<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> DerefMut for Flock<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> Flock<T> {
+    /// Obtain a/an flock.
+    ///
+    /// # Example
+    /// ```
+    /// # use std::fs::File;
+    /// # use nix::fcntl::{Flock, FlockArg};
+    /// fn do_stuff(file: File) -> nix::Result<()> {
+    ///     let lock = match Flock::lock(file, FlockArg::LockExclusive) {
+    ///         Ok(l) => l,
+    ///         Err((_, e)) => return Err(e),
+    ///     };
+    ///
+    ///     // Do stuff
+    ///
+    ///     Ok(())
+    /// } // File is unlocked once `lock` goes out of scope.
+    pub fn lock(t: T, args: FlockArg) -> std::result::Result<Self, (T, Errno)> {
```

Comment:
Given that the `FlockArg` type is only used for locking, should we remove those 2 `UnLock` variants from it?

Reply:
Well, I don't know why there is a variant `UnlockNonblock`, from the manual of FreeBSD, that `LOCK_NB` flag should be used with `LOCK_SH` or `LOCK_EX`

> A lock	is applied by specifying an operation argument that is one of	`LOCK_SH` or  `LOCK_EX`  with the optional	addition of  `LOCK_NB`.

The Linux manual does not mention what will happen if we use `LOCK_UN` and `LOCK_NB` together

Friendly cc @agrover, this type was added in #173

Reply:
BTW, if we are gonna change the `FlockArg` type, we can just re-define it, using `libc_bitflags!`:

```rs
libc_bitflags! {
    /// The `args` argument of [`Flock::lock()`].
    pub struct LockOperation: c_int {
        /// Place a shared lock
        LOCK_SH;
        /// Place an exclusive lock
        LOCK_EX;
        /// When included, the locking request will be nonblocking.
        LOCK_NB;
    }
}
```

And there is little point in keeping the `flock()` function, we can just remove it and call `libc::flock()` when implementing the `Flock` type

Reply:
> Well, I don't know why there is a variant `UnlockNonblock`, from the manual of FreeBSD, that `LOCK_NB` flag should be used with `LOCK_SH` or `LOCK_EX`
> 
> > A lock	is applied by specifying an operation argument that is one of	`LOCK_SH` or  `LOCK_EX`  with the optional	addition of  `LOCK_NB`.
> 
> The Linux manual does not mention what will happen if we use `LOCK_UN` and `LOCK_NB` together
> 
> Friendly cc @agrover, this type was added in #173

All it says is "To make a nonblocking request, include LOCK_NB (by ORing) with any of the above operations.", so it is a valid request. Having said that, there are no strong indications of a failure case for unlocking (one could argue that EWOULDBLOCK still applies), so it's probably a no-op in most cases (I think flock is a per-driver implementation).

> Given that the `FlockArg` type is only used for locking, should we remove those 2 `UnLock` variants from it?
> ...
> And there is little point in keeping the `flock()` function, we can just remove it and call `libc::flock()` when implementing the `Flock` type

If `nix` provides nice wrappers for things, shouldn't we leave it? We would also want to revert it to `pub` if we keep it, so people have the option (if they are doing something weird).

As an aside, I personally like the enum as-defined vice the libc bitflags, because it's more rust-like imo.

Reply:
> All it says is "To make a nonblocking request, include LOCK_NB (by ORing) with any of the above operations.", so it is a valid request. Having said that, there are no strong indications of a failure case for unlocking (one could argue that EWOULDBLOCK still applies), so it's probably a no-op in most cases (I think flock is a per-driver implementation).

Makes sence

> If nix provides nice wrappers for things, shouldn't we leave it? We would also want to revert it to pub if we keep it, so people have the option (if they are doing something weird).

For the `Flock::lock()` interface, we should remove that 2 unlock variants from it cause we don't want our users to do such things:

```rs
let lock = Flock::lock(file, FlockArg::Unlock).unwrap();
```

>  We would also want to revert it to pub if we keep it, so people have the option (if they are doing something weird).

IMHO, It isn't that good to provide 2 interfaces for the same thing. If people really want to do something weird, they can always choose the libc bindings (and nix has libc re-exported)


> As an aside, I personally like the enum as-defined vice the libc bitflags, because it's more rust-like imo.

I prefer the libc_flags way as it adheres to our [CONVENTIONS](https://github.com/nix-rust/nix/blob/master/CONVENTIONS.md#bitflags), and there is exactly a [TODO](https://github.com/nix-rust/nix/blob/dbdcfed86e18673c71b54d4c4cd7be289441843c/src/fcntl.rs#L599) for it

Reply:
> For the Flock::lock() interface, we should remove that 2 unlock variants...

Agreed.

> IMHO, It isn't that good to provide 2 interfaces for the same thing. If people really want to do something weird, they can always choose the libc bindings (and nix has libc re-exported)

Fair enough.

> I prefer the libc_flags way as it adheres to our [CONVENTIONS](https://github.com/nix-rust/nix/blob/master/CONVENTIONS.md#bitflags), and there is exactly a [TODO](https://github.com/nix-rust/nix/blob/dbdcfed86e18673c71b54d4c4cd7be289441843c/src/fcntl.rs#L599) for it

Preference and convention aren't inherently linked lol. Regardless, I would argue an enum type fits this use case just as well, if not better than the bitflags. Specifically, because all (sans `LOCK_NB`) the arguments to `flock()` are "constants that are intended as mutually exclusive arguments to parameters of functions", I think an enum is the best fit. Especially since non-blocking is the only change, I think having the extra enum variants is worth it given there's only 4 total.

Reply:
> constants that are intended as **mutually exclusive** arguments to parameters of functions

This makes sense to me, we should use an enum here

Reply:
And, please remove that TODO as it is no longer planned